### DISCUSSION TITLE
Практическое использование getArtistsTopTracks

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#19] [Помощь с алгоритмом] [answered]
Решил обновить до последней версии библиотеку.
И заодно заменить свою функцию getArtistsTopTracks на авторскую.

Какой пример из жизни использования новой функции getArtistsTopTracks? Не пойму пока как ее использовать.

Вот был у меня такой кусок кода, работал, с новой функцией не будет работать?
_TypeError: artists.reduce is not a function (line 206, file "!Library")_
Как передалать?

Здесь берется плейлист, оставляется 10 треков.
И далее по исполнителю каждого трека подтягивается топ его  треков.

```js
let sortOfSavedCopy = Selector.sliceCopy(sortOfSaved);
Filter.dedupArtists(sortOfSavedCopy);
Filter.removeArtists(sortOfSavedCopy, banArtistsMixOfMixes);
Selector.keepRandom(sortOfSavedCopy, 10);

let artistTopTracks = [];
sortOfSavedCopy.forEach((track) => {
    Combiner.push(artistTopTracks, Source.getArtistsTopTracks(track.artists[0].id));
});
Filter.removeTracks(artistTopTracks, banTracks);
Filter.removeTracks(artistTopTracks, recentTracks);
Filter.dedupArtists(artistTopTracks);
Selector.keepRandom(artistTopTracks, 10);
```

**Comment by Chimildic** [comment, marked as answer]
> TypeError: artists.reduce is not a function (line 206, file "!Library")

В вашем случае в `getArtistsTopTracks` передается одно `id` в типе строка. Согласно [описанию](https://chimildic.github.io/goofy/#/func?id=getartiststoptracks), функция принимает массив. 

Замените на следующий код:
```js
Source.getArtistsTopTracks([track.artists[0]])
```
Добавлены квадратные скобки `[]`. Они создадут массив с одним элементов - вашим `id`.

> Какой пример из жизни использования новой функции getArtistsTopTracks? 

Есть функции, которые возвращают массивы исполнителей. Например, [Source.getArtists](https://chimildic.github.io/goofy/#/func?id=getartists). Тогда можно получить топ 10 на каждого исполнителя. Например, для каждого отслеживаемого:
```js
let artists = Source.getArtists({ followed_include: true, });
let tracks = Source.getArtistsTopTracks(artists);
```

Аналогично для [Yandex.getArtists](https://chimildic.github.io/goofy/#/func?id=getartists-1). Позднее научится и модуль Lastfm.

**Reply by efremych** [reply]
_**Error: Not enough arguments (line 206, file "format_string.js")**_

```
    let sortOfSavedCopy = Selector.sliceCopy(sortOfSaved);
        Filter.dedupArtists(sortOfSavedCopy);
            Selector.keepRandom(sortOfSavedCopy, 10);
            
            let artistTopTracks = [];
              sortOfSavedCopy.forEach(track => {Combiner.push(artistTopTracks, Source.getArtistsTopTracks([track.artists[0].id]))});
                  Filter.dedupArtists(artistTopTracks);
                    Selector.keepRandom(artistTopTracks, 10);
                    
                    Console.log ('Кол-во', artistTopTracks.length);
```

P.S. На какую кнопку тут нажимать, чтобы красиво код вставлялся? :)

**Reply by Chimildic** [reply]
Ошибся немного, нужно убрать `.id`. Оно берется внутри функции само.
```js
Source.getArtistsTopTracks([track.artists[0]])
```

> На какую кнопку тут нажимать, чтобы красиво код вставлялся?

Тут нет автоформатирования. Я форматирую код в VS Code редакторе и вставляю сюда.
Сам код заключается в блок по три апострофа (кнопка ё). К первым трем добавляется `js`, чтобы подсветить синтаксис.
![2020-12-26_19-02-42](https://user-images.githubusercontent.com/40138097/103155070-57a0c380-47ad-11eb-8ba8-7d95ee99a297.png)

**Comment by efremych** [comment]
Получилось, спасибо!
Единсттвенное, так понял, в этой функции, что-то типа нескольких массивов на выходе.
Поэтому пришлось немного подкорректировать свою вставку по фиксации Источника трека.

**Было**
```js
responses[i].forEach((item) => item.source = 'Artist Top Tracks');
```
**Стало**
```js
for (let i = 0; i < responses.length; i++)
                {responses[i].forEach((item) => item.source = 'Artist Top Tracks')} 
```

В целом, вроде всё работает, обновление можно считать успешным :)
Почитаю, что там про переопределения, может воспользуюсь, спасибо!
